src/natcap/invest/validation.py

```diff
@@ -596,17 +601,60 @@ def check_spatial_overlap(spatial_filepaths_list,
     return None
 
 
+def timeout(func, *args, timeout=5, **kwargs):
+    """Stop a function after a given amount of time.
+
+    Args:
+        func (function): function to apply the timeout to
+        args: arguments to pass to the function
+        timeout (number): how many seconds to allow the function to run.
+            Defaults to 5.
+
+    Returns:
+        A string warning message if the thread completed in time and returned
+        warnings, ``None`` otherwise.
+
+    Raises:
+        ``RuntimeWarning`` if the thread does not complete in time.
+    """
+    # use a queue to share the return value from the file checking thread
+    # the target function puts the return value from `func` into shared memory
+    message_queue = queue.Queue()
+    def wrapper_func():
+        message_queue.put(func(*args, **kwargs))
+
+    thread = threading.Thread(target=wrapper_func)
+    LOGGER.info(f'Starting file checking thread with timeout={timeout}')
+    thread.start()
+    thread.join(timeout=timeout)
+
+    if thread.is_alive():
+        # first arg to `check_csv`, `check_raster`, `check_vector` is the path
+        warnings.warn(f'Validation of file {args[0]} timed out. If this file '
+            'is stored in a file streaming service, it may be taking a long '
+            'time to download. Try storing it locally instead.')
+        return None
+
+    else:
+        LOGGER.info('File checking thread completed.')
+        # get any warning messages returned from the thread
+        return message_queue.get()
+
+
+# opening a file could take a long time if it's in a file streaming service
+# to prevent the UI from hanging due to slow validation,
+# set a timeout for these functions.
 _VALIDATION_FUNCS = {
     'boolean': check_boolean,
-    'csv': check_csv,
+    'csv': functools.partial(timeout, check_csv),
     'file': check_file,
     'folder': check_directory,
     'directory': check_directory,
     'freestyle_string': check_freestyle_string,
     'number': check_number,
     'option_string': check_option_string,
-    'raster': check_raster,
-    'vector': check_vector,
+    'raster': functools.partial(timeout, check_raster),
+    'vector': functools.partial(timeout, check_vector),
     'other': None,  # Up to the user to define their validate()
```

Comment:
Neat use of `functools.partial`!

Do you think it'd be useful to also wrap `file`, `folder` and `directory` with the timeout function as well?  Since the idea of the new `timeout` function here is to have a timeout on file-based operations and these three also make OS calls to check if a file exists, I'm wondering if we're likely to get into a case where even checking to see if a file or folder exists on disk could take some time, like if someone is checking an NFS share or something like that.

As a separate note about the use of `functools.partial`, wrapping the timeout functions in this way unfortunately does not allow for the timeout to differ per use case.  Here are a few that come to mind that might benefit from different timeouts:

1. The Qt UI - here I think the default 5s timeout could work just fine ... we're just trying to prevent a feeling of unresponsiveness.
2. The workbench - we've previously talked about timeouts in the context of that it would be great to be able to basically guarantee fast validation ... so in that case, maybe we'd want validation to return within a much shorter timeframe.  In any case, I suspect it would be useful to be able to configure the timeout so that the workbench and/or server can set their own limits as needed.
3. Model execution - there are a number of model that call `validate` within the model's execute.  In this case it could be reasonable to permit a longer-running validation.

All that being said, perhaps this isn't the right place to address these timeout use cases, and the solution as stated _does_ address the issue at hand so maybe it'd be worth addressing this along with the other design issues in the validation API that came up on the slack thread yesterday.  What do you think @dcdenu4 @emlys ?

Reply:
It could definitely be useful to have a timeout for `file`, `folder`, and `directory` as well! I didn't include them at first because I'm pretty sure that doing `os.path.exists` on a file stream filepath wouldn't trigger it to download, but there are probably other situations where it could take a long time, so I'll add the timeouts there.

It should work to pass in a different time to `timeout`...
`functools.partial(timeout, check_csv, timeout=10)` --> `timeout(check_csv, *args, timeout=10, **kwargs)`, and then the thread target becomes `check_csv(*args, **kwargs)`. Changing the timeout this way does seem to work in the tests, so I'm pretty confident, but it's definitely not very intuitive how the args get passed around

Reply:
Ok great!  Yes, probably safest to add timeouts to all of the filesystem operations in case those take some time.

For the timeouts, it makes sense to pass the timeout argument around, and after some reflection, I think I might suggest that we wait to address that until we have a more concrete plan ironed out for a round of updates to validation.  What do you think?

Reply:
Agree that we can wait on adding that option until we plan to use it for those different cases 👍